labrad/fasttypes.c

```diff
@@ -0,0 +1,2521 @@
+                printf("dim_arr[%u]: %d\n", ui, cobj->dim_arr[ui]);
+                N *= cobj->dim_arr[ui];
+            }
+            if (!cobj->istype) {
+                for (ui=0; ui<N; ui++) {
+                    if (cobj->subtype == BOOL) {
+                        printf("arr[%u]: %d\n", ui, ((char *)cobj->arr)[ui]);
+                    }
+                    else if (cobj->subtype == UINT31 || cobj->subtype == INT32) {
+                        printf("arr[%u]: %d\n", ui, ((int *)cobj->arr)[ui]);
+                    }
+                    else if (cobj->subtype == UINT32) {
+                        printf("arr[%u]: %u\n", ui, ((unsigned int *)cobj->arr)[ui]);
+                    }
+                    else if (cobj->subtype == FLOAT64) {
+                        printf("arr[%u]: %g\n", ui, ((double *)cobj->arr)[ui]);
+                    }
+                    else if (cobj->subtype == COMPLEX128) {
+                        printf("arr[%u]: %g + %gj\n", ui, ((double *)cobj->arr)[2*ui], ((double *)cobj->arr)[2*ui+1]);
+                    }
+                }
+            }
+    }
+    if (!cobj->istype) {
+        switch (cobj->type) {
+            case TIME:
+                printf("secs: %lld\n", cobj->secs);
+                printf("usecs: %lld\n", cobj->usecs);
+        }
+    }
+    
+    // printf("\n");
+}
+
+void ft_print_type(COBJ *cobj) {
+    printf("type: ");
+    ft_print_type_text(cobj->type);
+    printf("\n");
+}
+
+void ft_print_subtype(COBJ *cobj) {
+    printf("subtype: ");
+    ft_print_type_text(cobj->subtype);
+    printf("\n");
+}
+
+void ft_print_type_text(int type) {
+   switch (type) {
+        case EMPTY:
+            printf("EMPTY");
+            break;
+        case ARBITRARY:
+            printf("ARBITRARY");
+            break;
+        case BOOL:
+            printf("BOOL");
+            break;
+        case UINT31:
+            printf("UINT31");
+            break;
+        case INT32:
+            printf("INT32");
+            break;
+        case UINT32:
+            printf("UINT32");
+            break;
+        case FLOAT64:
+            printf("FLOAT64");
+            break;
+        case COMPLEX128:
+            printf("COMPLEX128");
+            break;
+        case FLOAT64_SU:
+            printf("FLOAT64_SU");
+            break;
+        case COMPLEX128_SU:
+            printf("COMPLEX128_SU");
+            break;
+        case STRING:
+            printf("STRING");
+            break;
+        case CLUSTER:
+            printf("CLUSTER");
+            break;
+        case LIST:
+            printf("LIST");
+            break;
+        case NDARRAY:
+            printf("NDARRAY");
+            break;
+        case NDARRAY_SU:
+            printf("NDARRAY_SU");
+            break;
+        case TIME:
+            printf("TIME");
+            break;
+        case ERROR:
+            printf("ERROR");
+            break;
+    }
+}
+
+PyObject *ft_flatten_cobj(COBJ *cobj, COBJ *co
```

Comment:
Thanks Matthew, I'll get to work on these.

Regards,
Austin.

On Wed, May 6, 2015 at 6:16 PM, Matthew Neeley <notifications@github.com>
wrote:

> In labrad/fasttypes.c
> <https://github.com/martinisgroup/pylabrad/pull/88#discussion_r29817931>:
>
> > +        case NDARRAY:
> > +            printf("NDARRAY");
> > +            break;
> > +        case NDARRAY_SU:
> > +            printf("NDARRAY_SU");
> > +            break;
> > +        case TIME:
> > +            printf("TIME");
> > +            break;
> > +        case ERROR:
> > +            printf("ERROR");
> > +            break;
> > +    }
> > +}
> > +
> > +PyObject *ft_flatten_cobj(COBJ *cobj, COBJ *co
>
> The size of this buffer is never checked, which seems really dangerous for
> unflattening data coming from the network. We should probably 'traverse'
> the buffer first to verify that it has the right length and contains valid
> data of the specified type.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/martinisgroup/pylabrad/pull/88/files#r29817931>.
>